### PR TITLE
Detect and flag duplicate component names.

### LINKED ISSUE #185: Prevent Multiple Components With Same Name
Thought this was reported before but I guess not. Basically you can have multiple components with the same name. When your code gets run the Roku will choose one of the two components. Trying to diagnosis this can be a real time waster and should be easy enough to address.

### PR DESCRIPTION
**Pull request by TwitchBronBron**
Detect when there are multiple components with the same name, and flag each occurrence as an error.

Resolves #185

### CODE REVIEW

File: src/Program.spec.ts [hunk 1]
```diff
@@ -185,6 +186,46 @@ describe('Program', () => {
...
+        it('catches duplicate XML component names', async () => {
+            //add 2 components which both reference the same errored file
+            await program.addOrReplaceFile({ src: `${rootDir}/components/component1.xml`, dest: 'components/component1.xml' }, `
+                <?xml version="1.0" encoding="utf-8" ?>
+                <component name="Component1" extends="Scene">
+                </component>
+            `);
+            await program.addOrReplaceFile({ src: `${rootDir}/components/component2.xml`, dest: 'components/component2.xml' }, `
+                <?xml version="1.0" encoding="utf-8" ?>
+                <component name="Component1" extends="Scene" extends="Scene">
```
Comment (triwav) [inline src/Program.spec.ts:198]: Not sure if that's even valid xml with the double extends

File: src/files/XmlFile.ts [hunk 2]
```diff
@@ -176,6 +184,16 @@ export class XmlFile {
                             Position.create(lineIndex, idx),
                             Position.create(lineIndex, idx + 10)
                         );
+                        //calculate the rane of the component's name (if it exists)
```
Comment (triwav) [inline src/files/XmlFile.ts:187]: assume you meant name not rane?

Comment (elsassph) [reply src/files/XmlFile.ts:187]: More like range

File: src/Program.ts [hunk 3]
```diff
@@ -466,10 +468,52 @@ export class Program {
...
+                if (!map[lowerName]) {
+                    map[lowerName] = [];
+                }
+                map[lowerName].push(file);
+            }
+            return map;
+        }, {} as { [lowerComponentName: string]: XmlFile[] });
+
+        for (let componentName in componentsByName) {
+            let xmlFiles = componentsByName[componentName];
```
Comment (elsassph) [inline src/Program.ts:494]: Const ;)

File: src/files/XmlFile.ts [hunk 4]
```diff
@@ -176,6 +184,16 @@ export class XmlFile {
...
+                        let match = /(.*?name\s*=\s*(?:'|"))(.*?)('|")/.exec(this.lines[lineIndex]);
```
Comment (elsassph) [inline src/files/XmlFile.ts:188]: Const

File: src/Program.spec.ts (hunk 1 above)
Comment (TwitchBronBron) [reply src/Program.spec.ts:198]: oops. copy-paste error.

File: src/files/XmlFile.ts (hunk 2 above)
Comment (TwitchBronBron) [reply src/files/XmlFile.ts:187]: yeah, range. Good catch

File: src/files/XmlFile.ts (hunk 4 above)
Comment (TwitchBronBron) [reply src/files/XmlFile.ts:188]: yeah, might as well. This is a pretty big codebase, so haven't flipped the eslint switch yet to enforce this. I think I'll do this once we get your plugin PR merged, and the ternary PR merged.

**Comment by TwitchBronBron** [comment]
@elsassph @triwav I've addressed all of your comments now.

**Review by elsassph — APPROVED** [review]

**Pull request merged**